nixos/modules/tasks/auto-upgrade.nix

```
@@ -174,6 +252,233 @@ in
+                '';
+                type = lib.types.nullOr lib.types.int;
+              };
+
+              # Read-only options, defined here and for consumption by the end-user
+              options.desyncAge = lib.mkOption {
+                readOnly = true;
+                description = ''
+                  The age (in number of rebuilds) of this nixpkgs version.
+                '';
+                type = lib.types.int;
+              };
+
+              options.path = lib.mkOption {
+                readOnly = true;
+                description = ''
+                  The path to the last known-working version of nixpkgs that passes all
+                  `requireBuilds`.
+
+                  This option is read-only, and is the result of this package.
+                '';
+                type = lib.types.path;
+              };
+
+              options.pkgs = lib.mkOption {
+                readOnly = true;
+                description = ''
+                  The last known-working version of nixpkgs that passes all `requireBuilds`.
+
+                  This option is read-only, and is the result of this package.
+                '';
+                type = lib.types.pkgs;
+              };
+
+              # Internal options and configuration, to make this work
+              options.desyncUpgradeScript = lib.mkOption {
+                # Derivation called to upgrade the desync files.
+                # `$out` will be symlinked from `/var/lib/nixos/desyncs/${name}.nix by the `upgradeScript`
+                readOnly = true;
+                visible = false;
+                type = lib.types.package;
+              };
+
+              options.upgradeFailedScript = lib.mkOption {
+                # Derivation called when the latest upgrade failed.
+                # `$out` will be symlinked from `/var/lib/nixos/desyncs/${name}.nix by the `upgradeScript`
+                readOnly = true;
+                visible = false;
+                type = lib.types.package;
+              };
+
+              config =
+                let
+                  desync =
+                    if builtins.pathExists /var/lib/nixos/desyncs/${name}.nix then
+                      import /var/lib/nixos/desyncs/${name}.nix
+                    else
+                      {
+                        # If we have no desync saved, it means that any configuration that builds
+                        # would necessarily be with the default path.
+                        # This can happen eg. on the `nixos-rebuild` that adds the desync configuration
+                        # to a nixos config, that would most likely not happen during an auto-upgrade.
+                        desyncAge = 0;
+                        path = config.defaultPath;
+                      };
+                in
+                {
+                  desyncAge = desync.desyncAge;
+                  path = desync.path;
+                  pkgs = config.pkgsFor desync.path;
```

Should this be:

```suggestion
                  pkgs = config.pkgsFor config.path;
```

?

---

The goal here is to have an evaluated version of the packages that is from the latest working desync version. This version is recorded in the desync file, hence this choice :)